[Poker game step6] 승자 확인하기

1Consumption opened the pull request:
* 고민한 점
  1. 카드의 결과가 같다면(ex : 같은 투페어라면) 페어의 rank가 더 큰 플레이어를 우승 시켜야 하는데 이를 어떻게 할 지 고민했습니다. 이를 위해 Hand 클래스에 카드의 결과(스트레이트, 포카드, 트리플 등)에 대한 카드값(실제 값 : c6, d10, sK)을 가지는 Card 배열을 선언하였고, 이 값을 비교하여 승자를 정하게 했습니다.
  2. 또한 Card 객체에 Comparable 프로토콜을 적용하여 Card 내의 rank rawvalue를 꺼내오지 않고, Card 객체 자체를 비교할 수 있게 하였습니다.
  3. 포카드, 트리플의 경우 딕셔너리에 넣고 키를 카드로, 몇번 중복되었는지를 밸류로 가지면 좋을 것 같아서 [Card : Int]() 형태의 딕셔너리를 만들었습니다. 이 과정에서 Card를 hash 해야 했는데 Hashable 프로토콜을 적용하여 rank를 hash한 값을 hash 값으로 가지게 되어 딕셔너리의 key 역할을 할 수 있었습니다.
  4. 최대한 클래스 내부 변수는 클래스 내에서만 접근하도록 고민하였습니다.

* 아쉬운 점
  1. Hand 클래스가 너무 커진 것 같습니다.
  2. 알고리즘이 효율적이지 못하다고 생각됩니다.

* 궁금한 점
  1. 많은 클래스에 프로토콜을 채용하였는데, 이렇게 많이 채용해도 상관이 없는지 궁금합니다.

1Consumption committed:
너무 긴 메소드 이름 조정

1Consumption committed:
Card 클래스 내 Rank enum rawvalue type 변경

기존 String 타입에서 Int 타입으로 변경
랭크를 비교할 때 숫자끼리 비교하기 위함

1Consumption committed:
PokerGame 내 Stud enum 클래스 변경

segmented control의 index 값에 맞게 case 순서 바꿔줌

1Consumption committed:
승자 표시 이미지뷰 추가 및 메소드 이름 변경

승자를 표시하는 이미지뷰를 딜러에게 임시로 추가하였음
cardStackView와 승자 이미지뷰를 담는 resultStackView 추가
createCard(cardName: String) 메소드를 createImageView(imageName: String)으로 변경

1Consumption committed:
투명 이미지 추가

트로피 이미지가 없는 경우 투명 이미지로 대체

1Consumption committed:
Card 클래스 Equatable 프로토콜 수정

card의 rank를 바탕으로 같거나, 작거나, 큼을 비교

1Consumption committed:
Hand 클래스 내 resultPriority enum 생성

Hand에 있는 Card 배열에 맞게 결과값이 될 enum 생성

1Consumption committed:
Equatable 프로토콜을 Comparable 프로토콜로 변경

sort() 메소드를 사용할 때 어떤 기준으로 비교를할지 알려주기 위함

1Consumption committed:
뺄셈 동작 추가

rank의 rawValue를기준으로 뺄셈 함

1Consumption committed:
스트레이트, 포카드, 트리플 구분 알고리즘 추가

1Consumption committed:
pair 판단하는 메소드 작성

1Consumption committed:
Hand 클래스 내 checkStraight 메소드 테스트 함수 생성

1Consumption committed:
Hand 클래스 내 checkSameCard 메소드 테스트 함수 생성 및 수정

checkSameCard 메소드를 테스트하는 함수를 만들었고, 테스트를 통과하지 못하여 checkSameCard 메소드 및 checkStraight 메소드 수정

1Consumption committed:
Hand 클래스 내 checkPair 메소드 테스트 함수 생성 및 수정

checkPair 메소드를 테스트하는 함수를 만들었고, 테스트를 통과하지  못하여 checkPair 메소드 수정

1Consumption committed:
Card 클래스 내 Rank enum에 프로토콜 적용

Comparable 프로토콜을 적용하여 외부에서 rawValue에 접근하지 않고 rank를 비교할 수 있게 함

1Consumption committed:
Player 클래스에 카드의 결과를 받아오는 result() 메소드 생성

1Consumption committed:
페어가 3개 있으면 트리플로 인식하던 오류 수정

count를 초기화 하는 부분을 if 문에서 빼줌

1Consumption committed:
Hand 클래스 내 ResultPriority에 Comparable 프로토콜 추가

결과 비교를 위해 추가함

1Consumption committed:
PokerGame 클래스 내 winner 메소드 추가

현재는 핸즈가 같은 경우는 고려하지 않음.

1Consumption committed:
승자 표시 완료

Hand 클래스에 Comparable 프로토콜을 적용하여 Player가 가지고있는 Hand를 바탕으로 승자를 결정함

1Consumption committed:
readme 수정

추가한 기능 정리
스크린샷 첨부

1Consumption committed:
불필요한 코드 제거

1Consumption committed:
Player 클래스를 비교할 수 있도록 Comparable 프로토콜 채택

가지고있는 hand에서 resultCardInfo 를 바탕으로 비교

1Consumption committed:
Hand를 비교 조건 추가

같은 페어, 트리플, 포카드 등 조건이 같을 때 더 큰 카드를 가지고 있는 플레이어가 승자라고 판별 함

1Consumption committed:
winner 로직 변경

winner 를 선정할 때 Player 자체를 비교

1Consumption committed:
Card클래스에 Hashable 프로토콜 채용

딕셔너리에 넣기위해 해당 프로토콜 채용
rank를 hash 함

1Consumption committed:
Hand 내 checkSameCard 변경

핸즈가 같을 경우 rank가 높은 트리플, 포카드를 가진 플레이어가 승리하도록 수정

1Consumption committed:
Hand 내 checkStraight 변경

핸즈가 같을 경우 rank가 높은 스트레이트를 가진 플레이어가 승리하도록 수정

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -9,7 +9,16 @@
+    }
+    
     private var card: [Card] = [Card]()
+    var resultCardInfo: [Card] = [Card]()
```
card가 있는데 resultCardInfo 가 public 으로 필요한 이유를 모르겠습니다. 
private 으로 할 수는 없었나요?

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -9,7 +9,16 @@
+        }
+    }
+    
     private var card: [Card] = [Card]()
```
배열이라면 복수로 표현하는 게 좋겠습니다. 카드 하나라고 생각할 것 같습니다.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+        if checkSameCard(howMany: .triple, card: card.sorted()){
+            resultCardInfo.sort(by: >)
+            return .triple
+        }
```
`card.sorted()` 반복적으로 나오는데 매번 정렬을 하는 게 좋을까요? 
한 번만 정렬하고 반복해서 쓰는게 좋을까요?
그리고보니 Hand를 초기화하는 메소드가 없네요. 생성할 때부터 Hand.card는 정렬해서 갖고 있어도 되지 않을까요?
한 장씩 넣어주는군요. 점수를 계산할 때는 한꺼번에 갖고 있어도 되는거 아닐까요?

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+    }
+    
+    func checkStraight(card: [Card]) -> Bool{
+        var pre = card[0]
```
로컬 변수에서도 단수와 복수를 구분해주세요.
줄여서 pre 보다는 previousCard 처럼 좀 더 구체적이고 명확한 표현을 쓰세요.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+        var straightCount = 1
+        var preStraightCount = 0
+        for cardIndex in 1..<card.count{
+            if pre - card[cardIndex] == -1{
```
이렇게 값을 비교하는 것보다는 pre.isNext(card) 형태로 카드 객체가 비교하는 메소드를 갖고있고 그걸 사용하는 방식이 좋습니다.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+    
+    func checkSameCard(howMany same: ResultPriority, card: [Card]) -> Bool{
+        resultCardInfo.removeAll()
+        var howmany = [Card : Int]()
```
단어가 두 개면 Camel 케이스에서는 howMany로 써야겠네요.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+            } else{
+                howmany.updateValue(howmany[$0]! + 1, forKey: $0)
+            }
+        }
```
이 부분을 card.reduce()로 표현해볼 수 있을까요? 한 번 찾아보세요.

1Consumption replied:
reduce는 사용하지 못했지만 filter를 사용하여 바꿔봤습니다...!

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+            }
+            
+            pre = card[cardIndex]
+        }
```
checkPair 도 checkSameCard 와 같이 처리할 수 있지 않나요? 같은 카드가 2개 있는 경우도 같은거 아닌가요?
덧붙여서 check- 라는 표현보다 더 구체적인 동작을 나타내는 단어를 사용하세요. check-는 뭐를 체크한다는 것인지 알 수가 없습니다.

godrm commented on CardGameApp/CardGameApp/model/Hand.swift:
```diff
@@ -24,4 +33,127 @@ class Hand{
+            if lhs.card[index] != rhs.card[index]{
+                return false
+            }
+        }
```
이 정도 비교라면 card 배열 자체를 직접 비교해도 동일하지 않을까요?

1Consumption replied:
Hand를 비교할 때 card 배열 자체를 비교하면 각 card 내에서 제일 큰 카드에 의해 Hand가 비교되기 때문에 우선순위(스트레이트, 포카드, 트리플 등)를 반영하지 못한다는 점, 또한 우선순위가 같다면 그 이후 비교 조건을 적용하지 못한다는 점 때문에 다음과 같이 긴 비교문을 작성하게 되었습니다...

godrm replied:
그 중에 일부 비교는 배열이 같지 않은지를 비교하고 있어서 그랬습니다. 
여기서 카드를 비교한다는 게 어떤게 필요한 걸까요? 
이렇게 처리하는 건 가능할까요?
카드 배열 > 전체 점수 > 점수만 비교하기

godrm commented on CardGameApp/CardGameApp/model/Player.swift:
```diff
@@ -39,3 +43,13 @@ extension Player: CustomStringConvertible{
         return "Dealer"
     }
```
Player 클래스인데 내부에서 타입을 검사하는 비교문이 있네요. 
이렇게 하기 보다는 각각 구현하는 게 좋습니다. 그걸 다형성이라고 부릅니다.

1Consumption replied:
혹시 CustomStringConvertible 프로토콜을 채용한 extension을 말씀 하시는건가요?
CustomStringConvertible 프로토콜을 Dealer 내에서 못바꿔서 어쩔 수 없이 Player 내에서 비교를하고 한번에 처리를 해줬습니다.

godrm replied:
이런 식으로 처리하면 가능합니다. 비교문 대신 타입의 다형성을 활용하세요. 
```
class Player : CustomStringConvertible {
    var description: String {
        return "player"
    }
}

class Dealer : Player {
    override var description: String {
        return "dealer"
    }
}
```

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -15,8 +15,8 @@ class PokerGame{
     enum Stud: Int, CaseIterable{
-        case fiveCardStud = 5
         case sevenCardStud = 7
+        case fiveCardStud = 5
```
이거 순서는 왜 바꾸셨어요? 코드에 영향을 주나요?

1Consumption replied:
segmentedControl이 바꾼 값을 PokerGame에 setGameStyle을 통해 넘겨줍니다. 그런데 이 과정에서 Stud enum을 배열로 만들고, 바뀐 segmentedControl 인덱스에 의해 값을 넘겨주게 됩니다.
현재 view에 보여지는 stud segmentedControl 순서는 7Cards, 5Cards인데 7Cards를 골랐지만 fiveCardStud가 선택되기에 순서를 바꿔줬습니다.

godrm replied:
뷰의 순서 때문에 enum 순서를 바꾼 경우군요. 
그래서 index 값이 아니라 enum 값에 맞추는 게 좋습니다. 참고하세요.

godrm commented on CardGameApp/CardGameApp/model/PokerGame.swift:
```diff
@@ -68,6 +68,27 @@ class PokerGame{
+                    highest = $0
+                }
+            }
+        }
```
포커 게임이 players 배열을 탐색할 필요가 없습니다. 
`players.findHighest()` 호출하면 내부에서 탐색하고 비교해서 리턴해주면 되지 않나요?

godrm reviewed (commented):
다음 피드백을 확인해보세요.

1Consumption committed:
접근자 수정

1Consumption committed:
배열 이름 복수형으로 변경

1Consumption committed:
isSerial 메소드 추가

비교하는 카드의 랭크가 현재 카드의 바로 다음 랭크인지 확인하여 Bool 값을 리턴

1Consumption committed:
불분명한 변수, 동작을 명확한 이름이나 메소드로 대체함

1Consumption committed:
카멜케이스 적용

1Consumption committed:
checkPair 메소드제거

해당 메소드에서 해주는 일을 checkSameCard 메소드에 합침

1Consumption committed:
check로 시작하는 메소드 이름 변경

조금 더 구체적인 동작을 이름으로 변경해줌
getCardPriority() 메소드 내부 if else문 제거

1Consumption committed:
findHighest 메소드 위치 변경

PokerGame이 아닌 Players 객체에서 내부 변수에 접근하도록 변경

1Consumption committed:
Hand에 카드 추가 시 정렬 하게 해줌

PokerGame에서 Players 인스턴스에 모든 값이 들어오면 Players 내의 Player 인스턴스에 있는 Hand를 오름차순으로 소팅해줌

1Consumption commented:
피드백 반영하여 수정했습니다!!!!

godrm commented:
이 단계는 여기까지 머지하겠습니다. 위에 남긴 피드백을 확인해서 마무리합시다. 
수고하셨습니다.
